test/e2e/lib/helpers/broker_event_transformation_test_with_source_helper.go

```diff
@@ -0,0 +1,331 @@
+			t.Error(err)
+		}
+		if !out.Success {
+			// Log the output cloudauditlogssource pods.
+			if logs, err := client.LogsFor(client.Namespace, auditlogsName, lib.CloudAuditLogsSourceTypeMeta); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("cloudauditlogssource: %+v", logs)
+			}
+			// Log the output of the target job pods.
+			if logs, err := client.LogsFor(client.Namespace, targetName, lib.JobTypeMeta); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("job: %s\n", logs)
+			}
+			t.Fail()
+		}
+	}
+}
+
+func BrokerEventTransformationTestWithSchedulerSourceHelper(t *testing.T, client *lib.Client, authConfig lib.AuthConfig, brokerURL url.URL, brokerName string) {
+	data := "my test data"
+	targetName := "event-display"
+	sName := "scheduler-test"
+	// Create a target Job to receive the events.
+	lib.MakeSchedulerJobOrDie(client, data, targetName)
+	createTriggersAndKService(client, brokerName, targetName)
+
+	var url apis.URL = apis.URL(brokerURL)
+	// Just to make sure all resources are ready.
+	time.Sleep(5 * time.Second)
+
+	// Create the CloudSchedulerSource.
+	lib.MakeSchedulerOrDie(client, sName, data, targetName, authConfig.PubsubServiceAccount,
+		kngcptesting.WithCloudSchedulerSourceSinkURI(&url),
+	)
+
+	msg, err := client.WaitUntilJobDone(client.Namespace, targetName)
+	if err != nil {
+		t.Error(err)
+	}
+
+	t.Logf("Last termination message => %s", msg)
+	if msg != "" {
+		out := &lib.TargetOutput{}
+		if err := json.Unmarshal([]byte(msg), out); err != nil {
+			t.Error(err)
+		}
+		if !out.Success {
+			// Log the output of scheduler pods
+			if logs, err := client.LogsFor(client.Namespace, sName, lib.CloudSchedulerSourceTypeMeta); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("scheduler log: %+v", logs)
+			}
+
+			// Log the output of the target job pods
+			if logs, err := client.LogsFor(client.Namespace, targetName, lib.JobTypeMeta); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("addressable job: %+v", logs)
+			}
+			t.Fail()
+		}
+	}
+
+}
+
+func CreateKService(client *lib.Client) string {
+	kserviceName := helpers.AppendRandomString("kservice")
+	// Create the Knative Service.
+	kservice := resources.ReceiverKService(
+		kserviceName, client.Namespace)
+	client.CreateUnstructuredObjOrFail(kservice)
+	return kserviceName
+
+}
+
+func createTriggerWithKServiceSubscriber(client *lib.Client, brokerName, kserviceName string) {
+	dummyTriggerName := "dummy-broker-" + brokerName
+	client.Core.CreateTriggerOrFail(
+		dummyTriggerName,
+		eventingtestresources.WithBroker(brokerName),
+		eventingtestresources.WithAttributesTriggerFilter(
+			eventingv1alpha1.TriggerAnyFilter, eventingv1alpha1.TriggerAnyFilter,
+			map[string]interface{}{"type": "e2e-testing-dummy"}),
+		eventingtestresources.WithSubscriberServiceRefForTrigger(kserviceName),
+	)
+}
+
+func createTriggerWithTargetServiceSubscriber(client *lib.Client, brokerName, targetName string) {
```

Comment:
I don't see the difference between `createTriggerWithTargetServiceSubscriber` and `createTriggerWithKServiceSubscriber`, except for the trigger name

Reply:
How about `func reateTriggers(client *lib.Client, brokerName string, targets ...string)

Reply:
The extensions are also different. One use map[string]interface{}{"type": "e2e-testing-dummy"}) and the other use map[string]interface{}{"type": "e2e-testing-resp"}). 
Most of the code are already refactored into the function 
`func (client *Client) CreateTriggerOrFail(name string, options ...resources.TriggerOption) *v1alpha1.Trigger `

Reply:
You can save some duplicated code by doing`func reateTrigger(client *lib.Client, brokerName, triggerName, extensions)`, or just inline client.Core.CreateTriggerOrFail. But up to you.